experimental/devices/mfrc522/mfrc522.go

```diff
@@ -0,0 +1,767 @@
+		if err != nil {
+			return
+		}
+		select {
+		case _ = <-irqChannel:
+			break interruptLoop
+		case <-time.After(100 * time.Millisecond):
+			// do nothing
+		}
+	}
+	return
+}
+
+// AntiColl performs the collision check for different cards.
+func (r *Dev) AntiColl() (backData []byte, err error) {
+
+	err = r.devWrite(commands.BitFramingReg, 0x00)
+
+	backData, _, err = r.CardWrite(commands.PCD_TRANSCEIVE, []byte{commands.PICC_ANTICOLL, 0x20}[:])
+
+	if err != nil {
+		return
+	}
+
+	if len(backData) != 5 {
+		err = fmt.Errorf("Back data expected 5, actual %d", len(backData))
+		return
+	}
+
+	crc := byte(0)
+
+	for _, v := range backData[:4] {
+		crc = crc ^ v
+	}
+
+	if crc != backData[4] {
+		err = errors.New(fmt.Sprintf("CRC mismatch, expected %02x actual %02x", crc, backData[4]))
+	}
+
+	return
+}
+
+// CRC calculates the CRC of the data using the card chip.
+func (r *Dev) CRC(inData []byte) (res []byte, err error) {
+	res = []byte{0, 0}
+	err = r.clearBitmask(commands.DivIrqReg, 0x04)
+	if err != nil {
+		return
+	}
+	err = r.setBitmask(commands.FIFOLevelReg, 0x80)
+	if err != nil {
+		return
+	}
+	for _, v := range inData {
+		r.devWrite(commands.FIFODataReg, v)
+	}
+	err = r.devWrite(commands.CommandReg, commands.PCD_CALCCRC)
+	if err != nil {
+		return
+	}
+	for i := byte(0xFF); i > 0; i-- {
+		n, err1 := r.devRead(commands.DivIrqReg)
+		if err1 != nil {
+			err = err1
+			return
+		}
+		if n&0x04 > 0 {
+			break
+		}
+	}
+	lsb, err := r.devRead(commands.CRCResultRegL)
+	if err != nil {
+		return
+	}
+	res[0] = lsb
+
+	msb, err := r.devRead(commands.CRCResultRegM)
+	if err != nil {
+		return
+	}
+	res[1] = msb
+	return
+}
+
+// SelectTag selects the FOB device by device UUID.
+func (r *Dev) SelectTag(serial []byte) (blocks byte, err error) {
+	dataBuf := make([]byte, len(serial)+2)
+	dataBuf[0] = commands.PICC_SElECTTAG
+	dataBuf[1] = 0x70
+	copy(dataBuf[2:], serial)
+	crc, err := r.CRC(dataBuf)
+	if err != nil {
+		return
+	}
+	dataBuf = append(dataBuf, crc[0], crc[1])
+	backData, backLen, err := r.CardWrite(commands.PCD_TRANSCEIVE, dataBuf)
+	if err != nil {
+		return
+	}
+
+	if backLen == 0x18 {
+		blocks = backData[0]
+	} else {
+		blocks = 0
+	}
+	return
+}
+
+type AuthStatus byte
+
+const (
+	AuthOk AuthStatus = iota
+	AuthReadFailure
+	AuthFailure
+)
+
+func (r *Dev) auth(mode byte, blockAddress byte, sectorKey []byte, serial []byte) (authS AuthStatus, err error) {
+	buffer := make([]byte, 2)
+	buffer[0] = mode
+	buffer[1] = blockAddress
+	buffer = append(buffer, sectorKey...)
+	buffer = append(buffer, serial[:4]...)
+	_, _, err = r.CardWrite(commands.PCD_AUTHENT, buffer)
+	if err != nil {
+		authS = AuthReadFailure
+		return
+	}
+	n, err := r.devRead(commands.Status2Reg)
+	if err != nil {
+		return
+	}
+	if n&0x08 != 0 {
+		authS = AuthFailure
+	}
+	authS = AuthOk
+	return
+}
+
+// StopCrypto stops the crypto chip.
+func (r *Dev) StopCrypto() (err error) {
+	err = r.clearBitmask(commands.Status2Reg, 0x08)
```

Comment:
These constants should reference the datasheet page and section, in this case page 43, section 9.3.1.9. That's much more useful than named constants IMHO.

This can be done in a follow up.